### PR TITLE
change(web): updates mobile screen dimension detection code ⬅️

### PR DESCRIPTION
**Pull request by jahorton**
Based on #6787, as its user test issues motivated development of this PR.

So, it turns out there there are some very strange issues with how `window.innerWidth` updates on mobile devices, and these issues were behind the failed user tests.  It appears that the best approach for detecting screen width (what the mobile-form OSK uses) is actually `document.documentElement.clientWidth`.  (Minor reference: https://stackoverflow.com/a/54812656)

Relying upon that, along with some other shifts to how the mobile-form OSK computes its preferred height, appear to stabilize the base PR's issues.  That said, these are some notable shifts to how the calculations are done, and there may be noteworthy history behind why things were the way that they were, so I expect some scrunity in the review & in the user-test suite.

----

Alternatively, adding a 50ms timeout halfway during the rotation adjustments _seems_ to be enough to remedy the issue.  There's some very odd behavior that I noticed when stepping through this function via debugger:

https://github.com/keymanapp/keyman/blob/16643d7479b2dc751b4e8ccaa3f5627bc2570bda/web/source/kmwrotation.ts#L44-L67

Before `keyman.alignInputs()` is called, `window.innerWidth` is consistently "wrong" - it's notably larger than the value we'd expect it to hold.  _After_ it's called, when stepping through, `window.innerWidth` changes.  Unfortunately, that appears to be asynchronous, and a very short timeout isn't enough to alleviate the issue.  And there's nothing in that function that should directly affect the browser's management of `window.innerWidth`.

So, that "halfway during" note seen above would involve setting a 50ms timeout to run lines 47-60.  Possible to do, but that feels more "hacky" than the approach I decided to run with.  That _may_ just be my perception, though.

By the way, setting the rotation "resolve" timer to a longer period (i.e. further delaying the whole function shown above)  does _not_ remedy the issue.  Somehow, it's about that line - `keyman.alignInputs` - so far as I can tell.

# User Testing

SUITE_ROTATION

## Test environments

- GROUP_IOS_PHONE - any modern iOS phone should suffice
- GROUP_IOS_TABLET - any modern iOS tablet should suffice
- GROUP_OLD_IOS - try to find an iOS 12 or iOS 13 device to test with
- GROUP_ANDROID_PHONE - any modern Android phone should suffice
- GROUP_ANDROID_TABLET - any modern Android tablet should suffice
- GROUP_ANDROID_7 - run tests against early Android if possible (Android 7.0)

## Tests

- **TEST_OSK_PORTRAIT_LOAD**:

  1. Before loading the test page, have the device in a portrait orientation.
  2. Load the KeymanWeb testing page "Test unminified Keymanweb".
  3. Select the top text area in order to display the OSK.
  4. Report on the appearance of the OSK.  Does it seem reasonably positioned and sized?  No cropping issues?

- **TEST_OSK_LANDSCAPE_LOAD**:

  1. Before loading the test page, have the device in a landscape orientation.
  2. Load the KeymanWeb testing page "Test unminified Keymanweb".
  3. Select the top text area in order to display the OSK.
  4. Report on the appearance of the OSK.  Does it seem reasonably positioned and sized?  No cropping issues?

- **TEST_BROWSER_ROTATE_P-TO-L**:

  1. Before loading the test page, have the device in a portrait orientation.
  2. Load the KeymanWeb testing page "Test unminified Keymanweb".
  3. Select the top text area in order to display the OSK.
  4. Open the language menu, but do not change keyboards.
  5. Once loaded, rotate the device to a landscape orientation.
      - The test "passes" if the OSK rotates properly.
      - The test "fails" if the OSK is not visible after rotation.

- **TEST_BROWSER_ROTATE_L-TO-P**:

  1. Before loading the test page, have the device in a landscape orientation.
  2. Load the KeymanWeb testing page "Test unminified Keymanweb".
  3. Select the top text area in order to display the OSK.
  4. Open the language menu, but do not change keyboards.
  6. Once loaded, rotate the device to a portrait orientation.
      - The test "passes" if the OSK rotates properly.
      - The test "fails" if the OSK is not visible after rotation.

### CODE REVIEW

**Comment by bharanidharanj** [comment]
## SUITE_ROTATION: 

### GROUP_ANDROID_PHONE: - any modern Android phone should suffice

* **TEST_OSK_PORTRAIT_LOAD (PASSED):** Tested in the KeymanWeb testing page "Test unminified Keymanweb" (as per the test steps) and it is working as expected. (in my Android Mobile device -  Ver 11)



* **TEST_OSK_LANDSCAPE_LOAD (PASSED):** Tested in the KeymanWeb testing page "Test unminified Keymanweb" (as per the test steps) and it is working as expected. 




* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** Tested in the KeymanWeb testing page "Test unminified Keymanweb" (as per the test steps) and it is working as expected. 




* **TEST_BROWSER_ROTATE_L-TO-P (PASSED):** Tested in the KeymanWeb testing page "Test unminified Keymanweb" (as per the test steps) and it is working as expected.

**Comment by mcdurdin** [comment]
> Somehow, it's about that line - `keyman.alignInputs` - so far as I can tell.

So it sounds like `window.innerWidth` is internally invalidated, but refresh of the value is delayed. My suspicion is that the code in `keyman.alignInputs` is [forcing a DOM layout or reflow](https://gist.github.com/paulirish/5d52fb081b3570c81e3a), which internally also ends up validating `window.innerWidth`.

If this is the case, then when `alignInputs` goes away with #3030, we'll need to re-fix this. We should be able to do that by forcing a reflow using one of the triggers in the aforementioned link.

(Technically, this could be classed as a bug in Chrome I guess -- reading `window.innerWidth` should force a layout reflow post-rotation? Building a minimal repro might be painful though.)

BTW, that link above is a hugely worthwhile resource to wrap your brain around -- in terms of performance of the OSK in web in particular. Some of the notes may be slightly out of date but mostly very solid as far as I can tell.

**Comment by mcdurdin** [comment]
> change(web): updates mobile screen-dim detection code ⬅️

I didn't understand this PR title for a long time... what does this have to with screen dimming?

I find it helpful when approaching a new PR to have a short summary at the start of the initial comment stating what the purpose of the PR is. Just one or two sentences is good, introducing the purpose and approach. Something along the lines of "change the screen dimension calculations to use x.y.z instead of a.b.c, because a.b.c appears to be unreliable."

**Comment by mcdurdin** [comment]
What about this: https://stackoverflow.com/a/49383279/1836776?

**Comment by mcdurdin** [comment]
`orientationchange` is apparently deprecated: https://developer.mozilla.org/en-US/docs/Web/API/Window/orientationchange_event

See also:
* https://developer.mozilla.org/en-US/docs/Web/API/ScreenOrientation
* https://bugs.webkit.org/show_bug.cgi?id=170595
* https://github.com/gajus/orientationchangeend

**Comment by jahorton** [comment]
> `orientationchange` is apparently deprecated: https://developer.mozilla.org/en-US/docs/Web/API/Window/orientationchange_event
> 
> See also:
> 
>     * https://developer.mozilla.org/en-US/docs/Web/API/ScreenOrientation
> 
>     * https://bugs.webkit.org/show_bug.cgi?id=170595
> 
>     * https://github.com/gajus/orientationchangeend

And the suggested replacement... isn't even implemented on Safari yet.  I'll take the one that's fully cross-platform.

**Comment by mcdurdin** [comment]
> And the suggested replacement... isn't even implemented on Safari yet. I'll take the one that's fully cross-platform.

First draft of that spec was published 2012-05-22. Only 10 years ago. Given this is working around what is effectively a bug, we may have better results with multiple solutions to the issue. I'm not opposed to a big comment saying "do this for Safari which doesn't have the Orientation API" -- and using feature detection to get us an answer.

**Comment by bharanidharanj** [comment]
## SUITE_ROTATION: 

### GROUP_IOS_PHONE: - any modern iOS phone should suffice

* **TEST_OSK_PORTRAIT_LOAD (PASSED):** Tested this as per the instructions  in iOS 15.5 / iPhone 13 Pro Max Simulator  and it is working as expected. 




* **TEST_OSK_LANDSCAPE_LOAD (PASSED):** Tested this as per the instructions  in iOS 15.5 / iPhone 13 Pro Max Simulator  and it is working as expected. 




* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** Tested this as per the instructions  in iOS 15.5 / iPhone 13 Pro Max Simulator  and it is working as expected. 




* **TEST_BROWSER_ROTATE_L-TO-P (PASSED):** Tested this as per the instructions  in iOS 15.5 / iPhone 13 Pro Max Simulator  and it is working as expected. 




### GROUP_IOS_TABLET: - any modern iOS tablet should suffice

* **TEST_OSK_PORTRAIT_LOAD (PASSED):** Tested this as per the instructions  in iOS 15.5 / iPad 9th generation Simulator  and it is working as expected. 




* **TEST_OSK_LANDSCAPE_LOAD (PASSED):** Tested this as per the instructions  in iOS 15.5 / iPad 9th generation Simulator  and it is working as expected. 




* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** Tested this as per the instructions  in iOS 15.5 / iPad 9th generation Simulator  and it is working as expected. 




* **TEST_BROWSER_ROTATE_L-TO-P (PASSED):** Tested this as per the instructions  in iOS 15.5 / iPad 9th generation Simulator  and it is working as expected.

**Comment by bharanidharanj** [comment]
## SUITE_ROTATION: 

### GROUP_OLD_IOS: - try to find an iOS 12 or iOS 13 device to test with

* **TEST_OSK_PORTRAIT_LOAD (PASSED):** Tested as per the instructions in iOS 12.4 / iPhone 8 Simulator and it is working as expected. 




* **TEST_OSK_LANDSCAPE_LOAD (PASSED):** Tested as per the instructions in iOS 12.4 / iPhone 8 Simulator and it is working as expected. 





* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** Tested as per the instructions in iOS 12.4 / iPhone 8 Simulator and it is working as expected. 






* **TEST_BROWSER_ROTATE_L-TO-P (PASSED):** Tested as per the instructions in iOS 12.4 / iPhone 8 Simulator and it is working as expected.

**Comment by bharanidharanj** [comment]
## SUITE_ROTATION: 

### GROUP_ANDROID_5: - run tests against Android API 21 if possible (Android 5.0)

* **TEST_OSK_PORTRAIT_LOAD (BLOCKED):** Tested this API 21 / Android 5.0 emulator and I am getting an error message while trying to open the Keyman Web Testing Page. 



* **TEST_OSK_LANDSCAPE_LOAD (BLOCKED):**  Tested this API 21 / Android 5.0 emulator and I am getting an error message while trying to open the Keyman Web Testing Page. 
* **TEST_BROWSER_ROTATE_P-TO-L (BLOCKED):**  Tested this API 21 / Android 5.0 emulator and I am getting an error message while trying to open the Keyman Web Testing Page. 
* **TEST_BROWSER_ROTATE_L-TO-P (BLOCKED):**  Tested this API 21 / Android 5.0 emulator and I am getting an error message while trying to open the Keyman Web Testing Page.

**Comment by jahorton** [comment]
> ## SUITE_ROTATION:
> ### GROUP_ANDROID_5: - run tests against Android API 21 if possible (Android 5.0)

Huh. that's odd.  Haven't we run tests against the TeamCity-hosted Web test page artifacts before?  I wonder why it's giving us errors this time; it's clearly a TC thing not due to this PR's changes.

**Comment by jahorton** [comment]
> What about this: https://stackoverflow.com/a/49383279/1836776?

We already do that.

https://github.com/keymanapp/keyman/blob/16a63d84ea2c23aebc327b0b486cf34356245cc5/web/source/kmwrotation.ts#L91-L125

This block also happens to be where we'd set the potential `screenOrientation.onchange` event handler.  I guess it'd come first in the Android-block's `if`-`else` chain?  I mean, I can add it if we care that strongly about it at the moment.

------

A bit of an aside, but...

> > change(web): updates mobile screen-dim detection code ⬅️
> 
> I didn't understand this PR title for a long time... what does this have to with screen dimming?

So, that angle never even occurred to me.  My mind went "screen **dim**ension" - eh, just shorten it, it'll be _obvious_.  It took me a moment or two to figure out where the confusion came from!

**Comment by MakaraSok** [comment]
# Test Results

## SUITE_ROTATION: 

### GROUP_ANDROID_TABLET: - any modern Android tablet should suffice

* **TEST_OSK_PORTRAIT_LOAD (PASSED):** No obvious issue found.
* **TEST_OSK_LANDSCAPE_LOAD (PASSED):** No obvious issue found.
* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** The OSK stays visible after rotating.
* **TEST_BROWSER_ROTATE_L-TO-P (PASSED):** The OSK stays visible after rotating.

**Comment by MakaraSok** [comment]
SUITE_ROTATION:

GROUP_ANDROID_5: - run tests against Android API 21 if possible (Android 5.0)

TEST_OSK_PORTRAIT_LOAD (BLOCKED): I also experience this same thing.

image

**Comment by mcdurdin** [comment]
Re blocked tests @bharanidharanj, can we try against Android 6 or 7? (I am investigating options for supporting Android 5, but it looks like it is too old to work with new websites)

**Comment by jahorton** [comment]
> Re blocked tests @bharanidharanj, can we try against Android 6 or 7? (I am investigating options for supporting Android 5, but it looks like it is too old to work with new websites)

That does sound like the best approach we can take for the moment with our current testing setup.

**Comment by bharanidharanj** [comment]
> Re blocked tests @bharanidharanj, can we try against Android 6 or 7? (I am investigating options for supporting Android 5, but it looks like it is too old to work with new websites)

@mcdurdin @jahorton That's sounds good. I will try it in Android 6 emulator and will post my comment. Thanks.

**Comment by mcdurdin** [comment]
> @mcdurdin @jahorton I am getting the same issue in Android 6.0 emulator. But, it is working fine in Android 7.0 emulator.
> 
> So, Shall we use Android 7.0 emulator for future testing?

Here's what I think we should do:

1. For KeymanWeb testing, we test on Android 7.0 as our minimum.
2. For Keyman for Android testing, we continue to test on Android 5.0 as our minimum.

File: web/source/osk/anchoredOskView.ts [hunk 1]
```diff
@@ -150,17 +150,27 @@ namespace com.keyman.osk {
...
+      var oskHeightLandscapeView=Math.floor(Math.min(baseHeight, baseWidth)/2),
           height=oskHeightLandscapeView;
 
       if(device.formFactor == 'phone') {
-        var sx=Math.min(screen.height,screen.width),
-            sy=Math.max(screen.height,screen.width);
-
         if(keymanweb.util.portraitView())
-          height=Math.floor(Math.max(screen.availHeight,screen.availWidth)/3);
+          height=Math.floor(baseHeight/2.4);
```
Comment (mcdurdin) [inline web/source/osk/anchoredOskView.ts:171]: I don't understand why the divisor changed from `3` to `2.4` here?

File: web/source/osk/anchoredOskView.ts [hunk 2]
```diff
@@ -186,18 +196,18 @@ namespace com.keyman.osk {
...
-      } else if(device.OS == 'Android') {
-        try {
-          width=document.documentElement.clientWidth;
-        } catch(ex) {
+
+      width = document?.documentElement?.clientWidth;
+      if(typeof width == 'undefined') {
+        if(device.OS == 'iOS') {
+          // iOS does not interchange these values when the orientation changes!
+          //width = util.portraitView() ? screen.width : screen.height;
```
Comment (mcdurdin) [inline web/source/osk/anchoredOskView.ts:204]:
```suggestion
```

No need to leave commented code

File: web/source/osk/anchoredOskView.ts [hunk 3]
```diff
@@ -150,17 +150,27 @@ namespace com.keyman.osk {
         return keymanweb['getOskHeight']();
       }
 
-      var oskHeightLandscapeView=Math.floor(Math.min(screen.availHeight,screen.availWidth)/2),
+      let baseWidth  = document?.documentElement?.clientWidth;
```
Comment (mcdurdin) [inline web/source/osk/anchoredOskView.ts:153]: Could you add a comment explaining the rationale for the measurement source choices for future maintainers?

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Only minor things, marking as 'request changes' to keep the workflow moving :grin:

File: web/source/osk/anchoredOskView.ts (hunk 1 above)
Comment (jahorton) [reply web/source/osk/anchoredOskView.ts:171]: That's fair.

Note the addition of `baseHeight` earlier in the function, which is based on the height of the document element.  This is contained within the browser / WebView and will almost certainly be smaller than the _screen_, which originally was used to provide a base height.

In initial development and testing, keeping the value the same as before produced a very "short" keyboard, notably shrunken compared to before this branch's changes.  This new figure seems to approximate the old height pretty well.

This is also part of why I specified tablet targets for some of the user tests; I wanted to be sure the changes didn't cause uncomfortable side-effects there, as there would naturally be different scale-ratio between document and screen there.

Comment (mcdurdin) [reply web/source/osk/anchoredOskView.ts:171]: Cool bananas. Makes sense now. Can you document that above this line as a code comment for future maintainers?

**Comment by bharanidharanj** [comment]
## SUITE_ROTATION: 

### GROUP_ANDROID_TABLET: - any modern Android tablet should suffice

* **TEST_OSK_PORTRAIT_LOAD (PASSED):** Tested this in API 30 / 10.1 WXGA (Tablet) emualtor (Android 11.0) and it is working as expected. 

OSK_Portrait:

   

* **TEST_OSK_LANDSCAPE_LOAD (PASSED):** Tested this in API 30 / 10.1 WXGA (Tablet) emualtor (Android 11.0) and it is working as expected. 

OSK_Landscape:

   


* **TEST_BROWSER_ROTATE_P-TO-L (FAILED):** Tested this in API 30 / 10.1 WXGA (Tablet) emualtor (Android 11.0) and it seems that the OSK is not visible after rotation. 

before rotation:

   


after rotation:

   


* **TEST_BROWSER_ROTATE_L-TO-P (FAILED):** Tested this in API 30 / 10.1 WXGA (Tablet) emualtor (Android 11.0) and it seems that the OSK is not visible after rotation. 

before rotation:

   


after rotation:

**Comment by bharanidharanj** [comment]
## SUITE_ROTATION: 

### GROUP_IOS_PHONE: - any modern iOS phone should suffice

* **TEST_OSK_PORTRAIT_LOAD (PASSED):** Tested this as per the instructions in iOS 15.5 / iPhone 13 Pro Max Simulator and it is working as expected. 



* **TEST_OSK_LANDSCAPE_LOAD (PASSED):** Tested this as per the instructions in iOS 15.5 / iPhone 13 Pro Max Simulator and it is working as expected. 




* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** Tested this as per the instructions in iOS 15.5 / iPhone 13 Pro Max Simulator and it is working as expected. 

before rotation:




after rotation:



* **TEST_BROWSER_ROTATE_L-TO-P (PASSED):** Tested this as per the instructions in iOS 15.5 / iPhone 13 Pro Max Simulator and it is working as expected. 

before rotation:



after rotation:

**Comment by bharanidharanj** [comment]
## SUITE_ROTATION: 

### GROUP_IOS_TABLET: - any modern iOS tablet should suffice

* **TEST_OSK_PORTRAIT_LOAD (PASSED):** Tested this as per instructions in iOS 15.5 / iPad Pro (9.7 inch) Simulator and it is working as expected. 

   


* **TEST_OSK_LANDSCAPE_LOAD (PASSED):**  Tested this as per instructions in iOS 15.5 / iPad Pro (9.7 inch) Simulator and it is working as expected. 

   

* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):**  Tested this as per instructions in iOS 15.5 / iPad Pro (9.7 inch) Simulator and it is working as expected. 

before rotation:

   


after rotation:

   


* **TEST_BROWSER_ROTATE_L-TO-P (PASSED):**  Tested this as per instructions in iOS 15.5 / iPad Pro (9.7 inch) Simulator and it is working as expected. 

before rotation:

   

after rotation:

**Comment by bharanidharanj** [comment]
## SUITE_ROTATION: 

### GROUP_OLD_IOS: - try to find an iOS 12 or iOS 13 device to test with

* **TEST_OSK_PORTRAIT_LOAD (PASSED):** Tested this as per the instructions in iOS 13.7 / iPhone 11 Pro Max and it is working as expected.

   

* **TEST_OSK_LANDSCAPE_LOAD (PASSED):** TTested this as per the instructions in iOS 13.7 / iPhone 11 Pro Max and it is working as expected.

   

* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** Tested this as per the instructions in iOS 13.7 / iPhone 11 Pro Max and it is working as expected.

before rotation:

   

after rotation:

   


* **TEST_BROWSER_ROTATE_L-TO-P (PASSED):** Tested this as per the instructions in iOS 13.7 / iPhone 11 Pro Max and it is working as expected.

before rotation:

   

after rotation:

**Comment by bharanidharanj** [comment]
## SUITE_ROTATION: 

### GROUP_ANDROID_PHONE: - any modern Android phone should suffice

* **TEST_OSK_PORTRAIT_LOAD (PASSED):** Tested this as per the instructions in my Android Mobile device / Version 11.0 and it is working as expected. 

   


* **TEST_OSK_LANDSCAPE_LOAD (PASSED):** Tested this as per the instructions in my Android Mobile device / Version 11.0 and it is working as expected. 

   


* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** Tested this as per the instructions in my Android Mobile device / Version 11.0 and it is working as expected. 

before rotation:

   


after rotation:

   



* **TEST_BROWSER_ROTATE_L-TO-P (PASSED):** Tested this as per the instructions in my Android Mobile device / Version 11.0 and it is working as expected. 

before rotation:

   


after rotation:

**Comment by MakaraSok** [comment]
@keymanapp-test-bot retest SUITE_ROTATION GROUP_ANDROID_TABLET TEST_BROWSER_ROTATE_P-TO-L TEST_BROWSER_ROTATE_L-TO-P

I've observed that the OSK only stays visible when tested on an actual device. It quickly disappears when tested with an emulator, API 30.

**Comment by bharanidharanj** [comment]
# Test Results

## SUITE_ROTATION: 

### GROUP_ANDROID_TABLET: - any modern Android tablet should suffice

* **TEST_BROWSER_ROTATE_P-TO-L (PASSED):** As per Makara's comments it seems that this test case is passed in the actual device.
* **TEST_BROWSER_ROTATE_L-TO-P (PASSED):** As per Makara's comments it seems that this test case is passed in the actual device.

**Review by mcdurdin — APPROVED** [review]

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 16.0.43-alpha](https://keyman.com/downloads/releases/alpha/16.0.43)

**Pull request merged**